### PR TITLE
Add DynamicBuffer

### PR DESCRIPTION
**Pull request by nikwit**
## Proposed changes
Adds `DynamicBuffer`, a class that acts as a `std::vector<DataVector>` but allocates all  `DataVector`s in a single chunk. This will be needed to reduce allocations in `YlmSpherepack`.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: `const size_t` x 2   `noexcept` at end

Could you document this constructor? Honestly, I'm a bit unsure of what `outer_size` and `inner_size` are

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: These comments here apply generally below:
- don't need to mark anything in the body of a class as `inline`, that is implicit
- for value return types (e.g. `const_iterator`) I don't think you need to mark the resulting object as `const`. I.e. `typename std::vector<T>::const_iterator` is fine. If you want, you can also use `auto` as the return type there, or to be more careful, `decltype(auto)`. I'd personally switch all the `begin` and `end`s to return `auto` because it's fewer chars, but up to you :)

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: since this is just doing swaps member-wisely, is it needed?

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: no `_` on function names, even private ones (add `noexcept`)

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: What if instead of doing a type-swapping business on people, you always have people use the `DynamicBuffer` (removing the `Impl`). Then change `set_references` to:
```cpp
if constexpr (is_derived_of_vector_impl_v<T>) {
  for ...
}
```
and do a `Requires` of the same style to have the `outer_size, inner_size` constructor disappear and be replaced with a `size` constructor in the fundamental case?

This makes the type more uniform, and also means that we can forward declar `DynamicBuffer`. You can't forward declare a type alias

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DynamicBuffer.cpp]: after copying can you make sure the copied `DataVector`s no longer point into the old location? I see you reset the pointer, but you might still be able to access the memory (yay memory leaks and segfaults!) so it's best to explicitly check the location the copied `DataVector` points is different than the original

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DynamicBuffer.cpp]: isn't this just testing that `DataVector`s work not a property of the class you're adding or what am I missing?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (nikwit) [reply tests/Unit/DataStructures/Test_DynamicBuffer.cpp]: I suppose. I thought it was not entirely obvious that a non-owning DataVector would still do a size check but I guess there is no reason it would not. I can remove it.

File: src/DataStructures/DynamicBuffer.hpp
Comment (nikwit) [reply src/DataStructures/DynamicBuffer.hpp]: ok, I have done that. I have left the constructor with two sizes, where there is an error thrown when the size is not 1 in the fundemental case. This should allow users to call it like `DynamicBuffer(outer_size, get_size(data_vector_or_double))` without needing `constexpr if`.

It is however now hard to deduce the correct type for the `buffer_`. I would need a type traits map

(`DataVector`, `double`) -> `double`,
(`ComplexDataVector`, `std::complex<double>`) -> `std::complex<double>`.

Unfortunately I think `get_fundamental_type` maps `std::complex<double>` to `double`.

This can probably done quite easliy with SFINAE and `tt::is_complex_or_fundamental_v` but maybe it is unnecessary and we can leave it like it is. If we leave it to be working for `double` types only I can include a static assert that warns the user that this is not implemented.

File: src/DataStructures/DynamicBuffer.hpp
Comment (nikwit) [reply src/DataStructures/DynamicBuffer.hpp]: `std::swap` uses the move constructor so we need one or the other. I changed it to the move constructor now.

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: Can't the move semantics be default? member-wise move should be fine, all pointers, etc. will be perserved. I think you can just do `DynamicBuffer(DynamicBuffer&& other) noexcept = default;`

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: I think this is missing `&&` and should be default. You should also write an actual copy-assignment operator (I agree, it's rather similar to the copy constructor, but you can check for self-assignment there too)

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: could you rename `inner` and `outer` to something like `number_of_grid_points` and `number_of_vectors`? I don't think anyone will really know what `inner` and `outer` means (I don't...)

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: you're missing the `!` at the end here for this to be a doxygen comment

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: we prefer data members be after private member functions in classes

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (nilsdeppe) [reply tests/Unit/DataStructures/Test_DynamicBuffer.cpp]: I would vote to remove it because it's not testing any assertion in the code you actually added, you're re-testing other code :)

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DynamicBuffer.cpp]: can you verify this for _all_ `DataVector`s. It's easy to write code that works for only the first or last element but fails for things in the middle :) (I don't think that's the case here, but it's good to be sure)

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp:69]: Can you make this buffer serializable? (add a `pup` + test) We basically need everything to be serializable and it's annoying to discover something is missing a feature like `pup` and then having to add it later

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Looks good, you can squash, including any additional changes

File: src/DataStructures/DynamicBuffer.hpp
Comment (nikwit) [reply src/DataStructures/DynamicBuffer.hpp]: yis, I think you are right. I thought it would default to copying for non-owning `DataVector`s (as the move assignment does) but it appears it just moves pointers.

File: src/DataStructures/DynamicBuffer.hpp
Comment (nikwit) [reply src/DataStructures/DynamicBuffer.hpp]: This is the heart of the copy-and-swap idiom. By taking the parameter by value it acts as both the copy assignment operator and the move assignment operator. When called with an lvalue, the copy is made in the parameter list using the copy constructor, when called with an rvalue it will be move constructed. `std::swap` then uses the move constructor to swap the two objects around.

 Even nicer, it makes self-assignment checks redundant and (not that we care) guarantees strong exception safety.

**Comment by nikwit** [comment]
squashed, thank you for the review :)

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [reply src/DataStructures/DynamicBuffer.hpp]: How does this make self-assignment checks redundant? You perform an entire copy when you self-assign, or how does the copy constructor call magically get elided? I'd argue it also doesn't guarantee exception safety because the copy constructor doesn't guarantee that. That is, if I do `a = b;` this can throw because the copy constructor can throw, so nothing gained.

My general view on this is it's trying to be "too clever" where the simpler, standard code is easier to understand and has fewer nuances. The general advice is to _not_ do this when there's heap-allocated memory and write an actual copy and move assignment operators, so I'd prefer we do that.

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: does this actually do anything? I thought `data_vectors_` should be empty in this case so why bother sending it?

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [reply src/DataStructures/DynamicBuffer.hpp]: @wthrowe do you have any insight here?

Comment (wthrowe) [reply src/DataStructures/DynamicBuffer.hpp]: The posted code doesn't work.  If you try to assign the class you'll overflow the stack because swap performs assignments.

Comment (nikwit) [reply src/DataStructures/DynamicBuffer.hpp]: @nilsdeppe yes it will still make the copy. It is my understanding that self assignment checks are there to prevent proper ill-defined programs (e.g. when deleting the resource and then assigning to it), not to optimize away the copy. Self-assignment should be extremely rare and performing the check every call introduces its own cost.
Of course it can also still throw, it guarantees strong exception safety because all resources will be left intact. It does not guarantee no-throw as it necessarily allocates. This doesn't matter for spectre anyway it was just a side note.

@wthrowe you are right. The unit test did not catch that because I had naively assumed that `DynamicBuffer<T> dynamic_buffer_assigned = *dynamic_buffer;` would call the copy assignment operator but it just calls the copy constructor. For this to work properly I should not have removed the custom swap method or introduced a defaulted move assignment operator. 

I have changed it to a straight-forward implementation now. However, I do think that in general using very common design patterns such as the copy-and-swap idiom should not be discouraged. They are recommended pretty much anywhere you look when it comes to managing resources exactly because all nuances have been understood very well and because directly implementing all constructors has its own pitfalls.

File: src/DataStructures/DynamicBuffer.hpp
Comment (nikwit) [reply src/DataStructures/DynamicBuffer.hpp]: in case of fundamental types, all the data is saved in `data_vectors_` and the buffer is empty, so that all the iterators and similar still work.

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [reply src/DataStructures/DynamicBuffer.hpp]: - Yes, in general, we do have some places where self-assignment is done frequently, e.g. with the DataBox. In generic code it's not always all that obvious how frequently self-assignment will be done :Shrug: Of course, an if-statement unless evaluated in a tight loop is not exactly a huge cost in general either. Much lower than the dozens to hundreds of memory allocations this class saves :)
- Okay...
- It seems like in this case the idiom caused more problems than just implementing copy assignment, so it's not a very convincing case that this idiom is something that should be adopted :) As I mentioned before, I don't remember this idiom being the recommended one in any of the books I've read, but maybe I'm just not remembering. It seems like the place where this idiom provides a benefit is when a custom copy and move are required, but if only a custom copy is needed it seems a lot less useful, especially if a custom swap function needs to be written.

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DynamicBuffer.cpp:83]: [picking a line at random] we have functions `test_copy_semantics` and `test_move_semantics`. These assume an `operator==` is implemented, which is generally what is recommended if you can copy a class. I'd suggest considering using those functions to be sure copy and move are tested correctly

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [reply src/DataStructures/DynamicBuffer.hpp]: right. I don't have a great name for `data_vectors_`, maybe just `data_` to make it more accurate when `T` is not a `DataVector`? :shrug: up to you

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: The general advice is to avoid having inline constructors and large functions. Do you think the constructor and pup should be defined not inline? I'm not sure if this actually matters with class templates or if it's just classes

Comment (nikwit) [reply src/DataStructures/DynamicBuffer.hpp]: I think with templated classes the only downside you might have is an increase in build time. Not sure if thats a valid argument but if you read the `if constexpr` branches separately the functions are only about 3 lines each. I dont know how stringent spectre is on this in general so I really cant say.

**Comment by nikwit** [comment]
added `operator==`, and `test_copy_semantics` + `test_move_semantics` in unit test.

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [reply src/DataStructures/DynamicBuffer.hpp]: It looks like in general we define these out-of-line, which is what I remember the advice being in Scott Meyer's book

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: can you check the `number_of_grid_points_` too? Part of that check is to make sure the `pup` methods are correct. Then you can also just do `lhs.data_ == rhs.data_`. You'll need to make this a friend function, though

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp]: Can you remove the extra blank line here please?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Comment by nikwit** [comment]
changed to out of line definitions

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsdeppe) [inline src/DataStructures/DynamicBuffer.hpp:79]: I think you need to `// NOLINT` this line because clang-tidy has a bug in it (or at least as far as I can tell it's a bug)

**Review by nilsdeppe — COMMENTED** [review]
LGTM, clang-tidy is unhappy, I've added a comment as to what to `NOLINT` in this case. Please squash that in immediately :)

File: src/DataStructures/DynamicBuffer.hpp
Comment (nikwit) [reply src/DataStructures/DynamicBuffer.hpp:79]: squashed, thank you for your review :)

**Comment by nikwit** [comment]
now the docs complained for some reason. sorry for spamming the queue, I'd cancel if I could

**Review by nilsdeppe — DISMISSED** [review]
I cancelled the old CI runs :)

**Comment by kidder** [comment]
@wthrowe do you want to do the second review for this?  Feel free to decline if you don't have time

File: src/DataStructures/DynamicBuffer.hpp
Comment (wthrowe) [inline src/DataStructures/DynamicBuffer.hpp:15]: \cond forward declaration.

File: src/DataStructures/DynamicBuffer.hpp
Comment (wthrowe) [inline src/DataStructures/DynamicBuffer.hpp:52]: `noexcept`.  (I think it's the default, but make it consistent with the previous line.)

File: src/DataStructures/DynamicBuffer.hpp
Comment (wthrowe) [inline src/DataStructures/DynamicBuffer.hpp]: `noexcept` and again for the remaining methods.

File: src/DataStructures/DynamicBuffer.hpp
Comment (wthrowe) [inline src/DataStructures/DynamicBuffer.hpp:38]: Include `<type_traits>`, `"GetFundamentalType.hpp"`.

File: src/DataStructures/DynamicBuffer.hpp
Comment (wthrowe) [inline src/DataStructures/DynamicBuffer.hpp:48]: Include `<cstddef>`.

File: src/DataStructures/DynamicBuffer.hpp
Comment (wthrowe) [inline src/DataStructures/DynamicBuffer.hpp]: Use new-style `// NOLINTNEXTLINE(google-runtime-references)`.

File: src/DataStructures/DynamicBuffer.cpp
Comment (wthrowe) [inline src/DataStructures/DynamicBuffer.cpp]: I think we recently merged a check that will error on doxygen stuff in cpp files.  Rebase and then remove these if necessary.

File: src/DataStructures/DynamicBuffer.cpp
Comment (wthrowe) [inline src/DataStructures/DynamicBuffer.cpp]: `is_vector_data_type`?

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (wthrowe) [inline tests/Unit/DataStructures/Test_DynamicBuffer.cpp]: `const` x2

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (wthrowe) [inline tests/Unit/DataStructures/Test_DynamicBuffer.cpp]: This doesn't do anything, and there's a fourth one above.

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (wthrowe) [inline tests/Unit/DataStructures/Test_DynamicBuffer.cpp]: I think all the checks in this test should be exact.

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (wthrowe) [inline tests/Unit/DataStructures/Test_DynamicBuffer.cpp]: This is the entire test, so it should have a more accurate name.

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (wthrowe) [inline tests/Unit/DataStructures/Test_DynamicBuffer.cpp:21]: Put everything (except, optionally, the SPECTRE_TEST_CASE) in an anonymous namespace.

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (wthrowe) [inline tests/Unit/DataStructures/Test_DynamicBuffer.cpp:15]: On a quick scan, you need at least `<cstddef>`, `<limits>`, `<memory>`, `<random>`, `<utility>`, `<vector>`, `"DataVector.hpp"`, `"Gsl.hpp"`, and wherever `tt::is_iterable_v` is from.

File: src/DataStructures/DynamicBuffer.cpp
Comment (wthrowe) [inline src/DataStructures/DynamicBuffer.cpp:24]: Include `"Error.hpp"`.

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (wthrowe) [inline tests/Unit/DataStructures/Test_DynamicBuffer.cpp]: Why are you doing this?

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/DataStructures/DynamicBuffer.hpp
Comment (nilsvu) [reply src/DataStructures/DynamicBuffer.hpp:52]: Sometimes I had gcc7 complain about these noexcept declarations. Since they are indeed the default, should we just omit them from the defaulted rule-of-five?

File: tests/Unit/DataStructures/Test_DynamicBuffer.cpp
Comment (nikwit) [reply tests/Unit/DataStructures/Test_DynamicBuffer.cpp]: I have removed it since we are already checking that the memory addresses are not equal for each component of the copied/moved buffers.

File: src/DataStructures/DynamicBuffer.hpp
Comment (wthrowe) [reply src/DataStructures/DynamicBuffer.hpp:52]: I am fine with either way.  If you've seen trouble, let's remove them instead.

File: src/DataStructures/DynamicBuffer.hpp
Comment (wthrowe) [reply src/DataStructures/DynamicBuffer.hpp:15]: Er, sorry, just `/// \cond`.

**Review by wthrowe — CHANGES_REQUESTED** [review]
One place where my instructions were unclear, and @nilsleiffischer's request above.  Squash, including the remaining changes.

**Comment by nikwit** [comment]
thank you for the reviews, I squashed it.

**Review by wthrowe — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**
